NeoML/src/Dnn/Dnn.cpp

```diff
@@ -497,43 +497,38 @@ void CDnn::ForceRebuild()
 	sourceLayers.SetSize( 0 );
 }
 
-CDnn* CDnn::CreateReferenceDnn(CRandom& random)
-{
-	NeoAssertMsg(referenceDnnRegistoror.referenceCounter != -1,
-		"resricted creating reference dnn if dnn is reference itself");
-	for(int i = 0; i < layers.Size(); ++i) {
-		auto* srcLayer = dynamic_cast<CSourceLayer*>(layers[i].Ptr());
-		if (srcLayer != nullptr) {
-			CheckArchitecture(srcLayer->GetBlob().Ptr() != nullptr, srcLayer->GetName(),
-				"initialize source blobs before creating reference dnns");
-		}
+CDnn* CDnn::CreateReferenceDnn(CRandom* random)
+{
+	CDnn* originalDnn = ( referenceDnnRegister.referenceCounter == -1 ) ? referenceDnnRegister.originalDnn : this;
+	originalDnn->reshape();
+	
+	CDnnReferenceRegister referenceRegister(originalDnn);
+	if(random == nullptr) {
+		referenceRegister.originRandom = originalDnn->random;
 	}
 
-	reshape();
-
-	CDnn* newDnn = new CDnn(random, mathEngine);
-	newDnn->DisableLearning();
-
-	for(int i = 0; i < layers.Size(); ++i) {
-		CPtr<CBaseLayer> copyLayer;
+	CRandom& newRandom = (random == nullptr) ? referenceRegister.originRandom : *random;
+	CDnn* newDnn = new CDnn(newRandom, mathEngine);
```

Comment:
Add comment, that result dnn would be allocated by new, and user should delete it